tao_compiler/mlir/xla/ral/context/quantized_gpu_kernel_impl.cc

```diff
@@ -0,0 +1,323 @@
+  }
+};
+
+template <class T>
+inline void hash_combine(std::size_t& seed, const T& v) {
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+}
+
+struct CublasLtBiasKeyHasher {
+  std::size_t operator()(const CublasLtBiasParamsKey& key) const {
+    std::size_t seed = std::hash<int>()(key.m);
+    hash_combine(seed, key.k);
+    hash_combine(seed, key.n);
+    hash_combine(seed, key.lda);
+    hash_combine(seed, key.ldb);
+    hash_combine(seed, key.ldc);
+    hash_combine(seed, key.transa);
+    hash_combine(seed, key.transb);
+    hash_combine(seed, key.bias);
+    return seed;
+  }
+};
+
+struct CublasLtBiasState : public Context::Resource {
+  std::mutex mu;
+  std::unordered_map<CublasLtBiasParamsKey, CublasLtBiasParams*,
+                     CublasLtBiasKeyHasher>
+      cache;
+};
+
+void runCublasLtBiasForward(ExecutionContext* ctx, bool transa, bool transb,
+                            int m, int n, int k, void const* A, int lda,
+                            void const* B, int ldb, void* C, int ldc,
+                            void const* alpha, void const* beta,
+                            void const* bias) {
+  std::string unique_name = "tao_ral.gpu.cublasLt_bias_" +
+                            tao::ral::TaoTypeNameHelper<int8_t>::Invoke();
+  auto state = ctx->getOrCreateResource<CublasLtBiasState>(
+      unique_name, []() { return new CublasLtBiasState; });
+  CublasLtBiasParamsKey key{m, k, n, lda, ldb, ldc, transa, transb, bias};
+  {
+    std::lock_guard<std::mutex> l(state->mu);
+
+    auto& cache = state->cache;
+    auto it = cache.find(key);
+    if (it == cache.end()) {
+      auto params_tmp = new CublasLtBiasParams(key);
+      cache.insert(std::make_pair(key, params_tmp));
+    }
+  }
+  auto params = state->cache[key];
+
+  auto status =
+      cublasLtMatmul(params->ltHandle, params->matmulDesc, alpha, A,
+                     params->Adesc, B, params->Bdesc, beta, C, params->Cdesc, C,
+                     params->Cdesc, NULL, NULL, 0, 0);
+  if (status != CUBLAS_STATUS_SUCCESS) {
+    ctx->signalError(Context::FAILURE, "Error to execute cublasLt.");
+  }
+}
+
+struct GemmScaleKey {
+  opaque_t input_scale = nullptr;
+  opaque_t weight_scale = nullptr;
+  opaque_t output_scale = nullptr;
+
+  bool operator==(const GemmScaleKey& rhs) const {
+    return input_scale == rhs.input_scale && weight_scale == rhs.weight_scale &&
+           output_scale == rhs.output_scale;
+  }
+};
+
+struct GemmScaleKeyHasher {
+  std::size_t operator()(const GemmScaleKey& key) const {
+    std::size_t seed = std::hash<intptr_t>()((intptr_t)key.input_scale);
+    hash_combine(seed, key.weight_scale);
+    hash_combine(seed, key.output_scale);
+    return seed;
+  }
+};
+
+struct GemmScaleState : public Context::Resource {
+  std::mutex mu;
+  std::unordered_map<GemmScaleKey, float, GemmScaleKeyHasher> cache;
+};
+
+template <int N, typename T>
+MemRefType<int8_t, N> ral_pdll_qgemm(
```

Comment:
please document the layout requirements for each input.

Reply:
Done